### PR TITLE
8186188: TableColumHeader: initial auto-size broken if has graphic

### PR DESCRIPTION
**Pull request by Maran23**
This PR fixes the issue that the initial column autosizing is wrong under some circumstances.
The following things will break the initial autosizing:
- Bold Column text (that is where I initially found this problem)
- Another font / font size
- Graphic
- Graphic Content Display (which can be styled via CSS)

The reason is actually quite simple: The CSS is not (yet) applied initially, we therefore ALWAYS take the default font into account + the graphic is not yet layouted as well.

~It was not so easy to write tests for this, also for me the `test_resizeColumnToFitContentHeader` is always failing locally. I don't know what happens here, but he seems to not find a (Stub?) `Font` for me.~
**EDIT: Found out the cause and fixed it. I will check if I can write more tests since it works now. :)**

The test I wrote now is checking if the css is applied after we triggered the autosize, which is what we would expect here since we measure text.

I also copied the `TableColumnHeaderTest` and rewrote the tests for `TreeTableView` as well, so we can catch any errors here as well since they both use different code (although it is technically the same - C&P errors can happen very easy).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8186188](https://bugs.openjdk.org/browse/JDK-8186188): TableColumHeader: initial auto-size broken if has graphic (**Bug** - P4)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Robert Lichtenberger](https://openjdk.org/census#rlichten) (@effad - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1405/head:pull/1405` \
`$ git checkout pull/1405`

Update a local copy of the PR: \
`$ git checkout pull/1405` \
`$ git pull https://git.openjdk.org/jfx.git pull/1405/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1405`

View PR using the GUI difftool: \
`$ git pr show -t 1405`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1405.diff">https://git.openjdk.org/jfx/pull/1405.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1405#issuecomment-2002162304)

### CODE REVIEW

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class TableColumnHeaderTreeTableTest {
+
+    private TableColumnHeader firstColumnHeader;
+    private TreeTableView<Person> tableView;
+    private StageLoader sl;
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java:64]: minor: this field should probably be named `stageLoader`

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+     * Resize the column header without modifications
+     */
+    @Test
+    public void test_resizeColumnToFitContent() {
+        TreeTableColumn column = tableView.getColumns().get(0);
+        double width = column.getWidth();
+        TableColumnHeaderShim.resizeColumnToFitContent(firstColumnHeader, -1);
+
+        assertEquals("Width must be the same",
+                width, column.getWidth(), 0.001);
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java:153]: maybe a constant, here and elsewhere
```
private static final double EPSILON = 0.001;
```
?

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubFontLoader.java [hunk 3]
```diff
@@ -48,8 +48,14 @@ public void loadFont(Font font) {
         StubFont nativeFont = new StubFont();
         nativeFont.font = font;
         String name = font.getName().trim().toLowerCase(Locale.ROOT);
-        if (name.equals("system") || name.equals("system regular")) {
+        if (name.equals("system regular")) {
             FontHelper.setNativeFont(font, nativeFont, font.getName(), "System", "Regular");
+        } else if (name.equals("system bold")) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubFontLoader.java:53]: does this mean this PR requires #1422 to be integrated first?

**Review by andy-goryachev-oracle — COMMENTED** [review]
the main question is whether this PR depends on #1422, or whether the unrelated changes need to be removed.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+import javafx.scene.control.TreeTableRow;
+import javafx.scene.control.TreeTableView;
+import javafx.scene.control.cell.TreeItemPropertyValueFactory;
+import javafx.scene.control.skin.TableColumnHeader;
+import javafx.scene.control.skin.TableColumnHeaderShim;
+import javafx.scene.input.MouseButton;
+import javafx.scene.input.MouseEvent;
+import javafx.scene.layout.HBox;
+import javafx.scene.text.Text;
+import org.junit.After;
```
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableColumnHeaderTreeTableTest.java:48]: I copied the other test class, but sure, is probably better to switch this to JUnit 5 then.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubFontLoader.java (hunk 3 above)
Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubFontLoader.java:53]: Yes, that would be easier for reviewing this. Once merged, I will update this branch. :)

**Comment by Maran23** [comment]
@effad Since you benchmarked this method in #1358, could you do that again with this changes?

**Comment by effad** [comment]
> @effad Since you benchmarked this method in #1358, could you do that again with this changes?

Sorry for the late reply, I just returned from easter holidays :-). I'll try to benchmark this change within the next days...

**Review by effad — APPROVED** [review]
As per request I tested the changes of this PR with the Benchmark found at https://gist.github.com/effad/9eebee0c1e86a8e605cb55ced9485dd4

Baseline: JFX 23-internal+0-2024-04-02-130613 average run time: 975
After `gh pr checkout 1405`: JFX 23-internal+0-2024-04-02-130613 average run time: 1026

So we have a measurable (~5%) but (IMO) neglectable slowdown. Compared to JFX 21.0.2+5 (2460) we are still more than twice as fast.

**Comment by Maran23** [comment]
@effad Thank you, much appreciated! Good to see that this does not cause any major performance problems.

**Review by andy-goryachev-oracle — APPROVED** [review]
Works as expected with the code in the ticket, no issues found with the monkey tester.
Thanks!

**Pull request closed without merging**